aptos-move/block-executor/src/executor.rs

```diff
@@ -159,37 +164,47 @@ where
-        versioned_cache: &MVHashMap<T::Key, T::Value, ExecutableTestType>,
+        versioned_cache: &MVHashMap<T::Key, T::Value, X>,
         scheduler: &Scheduler,
     ) -> SchedulerTask {
         use MVDataError::*;
         use MVDataOutput::*;
 
         let _timer = TASK_VALIDATE_SECONDS.start_timer();
         let (idx_to_validate, incarnation) = version_to_validate;
-        let read_set = last_input_output
-            .read_set(idx_to_validate)
-            .expect("Prior read-set must be recorded");
-
-        let valid = read_set.iter().all(|r| {
-            match versioned_cache.fetch_data(r.path(), idx_to_validate) {
-                Ok(Versioned(version, _)) => r.validate_version(version),
-                Ok(Resolved(value)) => r.validate_resolved(value),
+        let inputs = last_input_output.inputs(idx_to_validate);
+
+        let mut valid = inputs.reads().iter().all(|(k, r)| {
+            match versioned_cache.fetch_data(k, idx_to_validate) {
+                Ok(Versioned(version, _)) => *r == ReadDescriptor::Version(version.0, version.1),
+                Ok(Resolved(value)) => *r == ReadDescriptor::Resolved(value),
                 // Dependency implies a validation failure, and if the original read were to
                 // observe an unresolved delta, it would set the aggregator base value in the
                 // multi-versioned data-structure, resolve, and record the resolved value.
                 Err(Dependency(_)) | Err(Unresolved(_)) => false,
-                Err(NotFound) => r.validate_storage(),
+                Err(NotFound) => *r == ReadDescriptor::Storage,
                 // We successfully validate when read (again) results in a delta application
                 // failure. If the failure is speculative, a later validation will fail due to
                 // a read without this error. However, if the failure is real, passing
                 // validation here allows to avoid infinitely looping and instead panic when
                 // materializing deltas as writes in the final output preparation state. Panic
                 // is also preferrable as it allows testing for this scenario.
-                Err(DeltaApplicationFailure) => r.validate_delta_application_failure(),
+                Err(DeltaApplicationFailure) => *r == ReadDescriptor::DeltaApplicationFailure,
             }
         });
 
+        valid = valid
+            && inputs.executables().iter().all(|(k, r)| {
+                // TODO: don't fail code validation on dependency, compare hash after
+                // the dependency has finished execution.
+                match versioned_cache.fetch_code(k, idx_to_validate) {
+                    Ok(MVCodeOutput::Executable((_, desc))) => *r == desc,
+                    Ok(MVCodeOutput::Module(_))
+                    | Err(MVCodeError::NotFound)
+                    | Err(MVCodeError::Dependency(_)) => false,
```

Comment:
I guess the Dependency part will come later with dynamic linking?

Reply:
Currently we register a read on the code path only when we get the executable (so if we get "module", we should create executable, put it, and keep reading until we read an executable), and if during validation we observe a "module", it does mean corresponding executable hasn't been cached, so the "read behavior" changed at that path as far as validation logic goes.